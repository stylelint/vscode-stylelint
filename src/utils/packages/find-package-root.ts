import fs from 'fs/promises';
import path from 'path';

/**
 * Walks up the file tree from the given path until it finds a directory
 * containing a file named `package.json`. Resolves to `undefined` if no such
 * directory is found.
 * @param startPath The path to start from.
 * @param rootFile The file to use to determine when the project root has been
 * reached. Defaults to `package.json`.
 */
export async function findPackageRoot(
	startPath: string,
	rootFile = 'package.json',
): Promise<string | undefined> {
	let currentDirectory = startPath;

	while (true) {
		const manifestPath = path.join(currentDirectory, rootFile);

		try {
			const stat = await fs.stat(manifestPath);

			if (stat.isFile()) {
				return currentDirectory;
			}

			const parent = path.dirname(currentDirectory);

			if (!path.relative(parent, currentDirectory)) {
				return undefined;
			}

			currentDirectory = parent;
		} catch (error) {
			if (
				(error as { code?: string }).code === 'ENOENT' ||
				(error as { code?: string }).code === 'ENOTDIR'
			) {
				const parent = path.dirname(currentDirectory);

				if (!path.relative(parent, currentDirectory)) {
					return undefined;
				}

				currentDirectory = parent;
			} else {
				throw error;
			}
		}
	}
}
