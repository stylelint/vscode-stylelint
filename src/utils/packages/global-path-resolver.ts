import os from 'os';
import path from 'path';
import type winston from 'winston';
import { runProcessFindLine } from '../processes';
import { PackageManager } from './types';

/**
 * Resolves the global `node_modules` path for different package managers.
 */
export class GlobalPathResolver {
	/**
	 * The logger to use for tracing resolution.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * The cache of resolved paths.
	 */
	#cache: { [key in PackageManager]: string | undefined } = {
		yarn: undefined,
		npm: undefined,
		pnpm: undefined,
	};

	/**
	 * Whether or not the current platform is Windows.
	 */
	#isWindows: boolean;

	/**
	 * The resolvers by package manager.
	 */
	#resolvers: { [key in PackageManager]: () => Promise<string | undefined> } = {
		yarn: this.#yarn.bind(this),
		npm: this.#npm.bind(this),
		pnpm: this.#pnpm.bind(this),
	};

	/**
	 * Instantiates a new global path resolver.
	 * @param logger The logger to use for tracing resolution.
	 */
	constructor(logger?: winston.Logger) {
		this.#logger = logger;
		this.#isWindows = os.platform() === 'win32';
	}

	/**
	 * Resolves the global `node_modules` path for Yarn.
	 *
	 * Note: Only Yarn 1.x is supported. Yarn 2.x and higher have removed
	 * support for globally installed packages.
	 */
	async #yarn(): Promise<string | undefined> {
		const tryParseLog = (line: string): { type: string; data: string } | undefined => {
			try {
				return JSON.parse(line) as { type: string; data: string };
			} catch {
				return undefined;
			}
		};

		const yarnGlobalPath = await runProcessFindLine(
			'yarn',
			['global', 'dir', '--json'],
			this.#isWindows ? { shell: true } : undefined,
			(line) => {
				const log = tryParseLog(line);

				if (!log || log.type !== 'log' || !log.data) {
					return undefined;
				}

				const globalPath = path.join(log.data, 'node_modules');

				this.#logger?.debug('Yarn returned global node_modules path.', {
					path: globalPath,
				});

				return globalPath;
			},
		);

		if (!yarnGlobalPath) {
			this.#logger?.warn('"yarn global dir --json" did not return a path.');

			return undefined;
		}

		return yarnGlobalPath;
	}

	/**
	 * Resolves the global `node_modules` path for npm.
	 */
	async #npm(): Promise<string | undefined> {
		const npmGlobalPath = await runProcessFindLine(
			'npm',
			['config', 'get', 'prefix'],
			this.#isWindows ? { shell: true } : undefined,
			(line) => {
				const trimmed = line.trim();

				if (!trimmed) {
					return undefined;
				}

				const globalPath = this.#isWindows
					? path.join(trimmed, 'node_modules')
					: path.join(trimmed, 'lib/node_modules');

				this.#logger?.debug('npm returned global node_modules path.', {
					path: globalPath,
				});

				return globalPath;
			},
		);

		if (!npmGlobalPath) {
			this.#logger?.warn('"npm config get prefix" did not return a path.');

			return undefined;
		}

		return npmGlobalPath;
	}

	/**
	 * Resolves the global `node_modules` path for pnpm.
	 */
	async #pnpm(): Promise<string | undefined> {
		const pnpmGlobalPath = await runProcessFindLine(
			'pnpm',
			['root', '-g'],
			this.#isWindows ? { shell: true } : undefined,
			(line) => {
				const trimmed = line.trim();

				if (!trimmed) {
					return undefined;
				}

				this.#logger?.debug('pnpm returned global node_modules path.', {
					path: trimmed,
				});

				return trimmed;
			},
		);

		if (!pnpmGlobalPath) {
			this.#logger?.warn('"pnpm root -g" did not return a path.');

			return undefined;
		}

		return pnpmGlobalPath;
	}

	/**
	 * Attempts to resolve the global `node_modules` path for the given package
	 * manager.
	 *
	 * On a successful resolution, the method returns a promise that resolves to the
	 * package manager's global `node_modules` path. Paths are cached in the
	 * resolver on the first successful resolution.
	 *
	 * When a path cannot be resolved, the promise resolves to `undefined`.
	 *
	 * @example
	 * ```js
	 * const resolver = getGlobalPathResolver();
	 * const yarnGlobalPath = await resolver.resolve(
	 *   'yarn',
	 *   message => connection && connection.tracer.log(message)
	 * );
	 * ```
	 * @param packageManager The package manager to resolve the path for.
	 */
	async resolve(packageManager: PackageManager): Promise<string | undefined> {
		const cached = this.#cache[packageManager];

		if (cached) {
			return cached;
		}

		const resolver = this.#resolvers[packageManager];

		if (!resolver) {
			this.#logger?.warn('Unsupported package manager.', { packageManager });

			return undefined;
		}

		try {
			const globalPath = await resolver();

			if (globalPath) {
				this.#cache[packageManager] = globalPath;
			}

			return globalPath;
		} catch (error) {
			this.#logger?.warn('Failed to resolve global node_modules path.', {
				packageManager,
				error,
			});

			return undefined;
		}
	}
}
