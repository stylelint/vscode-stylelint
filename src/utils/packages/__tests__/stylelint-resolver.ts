jest.mock('vscode-languageserver/node');
jest.mock('../global-path-resolver');
jest.mock('../find-package-root');
jest.mock('path');
jest.mock('fs/promises', () => jest.createMockFromModule('fs/promises'));
jest.mock('module');

import path from 'path';
import fs from 'fs/promises';
import module from 'module';
import type winston from 'winston';
import { Connection, Files } from 'vscode-languageserver/node';
import type { TextDocument } from 'vscode-languageserver-textdocument';
import { GlobalPathResolver } from '../global-path-resolver';
import { findPackageRoot } from '../find-package-root';
import { StylelintResolver } from '../stylelint-resolver';
import type { Stats } from 'fs';
import type { PackageManager } from '../index';

const mockedPath = path as tests.mocks.PathModule;
const mockedFS = fs as jest.Mocked<typeof fs>;
const mockedModule = module as jest.Mocked<typeof module>;
const mockedFiles = Files as tests.mocks.VSCodeLanguageServerModule.Node['Files'];
const mockedGlobalPathResolver = GlobalPathResolver as jest.Mock<GlobalPathResolver>;
const mockedFindPackageRoot = findPackageRoot as jest.MockedFunction<typeof findPackageRoot>;

let mockCWD: string | undefined = mockedPath.join('/fake', 'cwd');
let mockPnPVersion: string | undefined = undefined;

jest.mock('../../documents', () => ({
	getWorkspaceFolder: jest.fn(async () => mockCWD),
}));

jest.mock('process', () => ({
	versions: {
		get pnp() {
			return mockPnPVersion;
		},
	},
}));

const createMockConnection = () =>
	({
		console: { error: jest.fn() },
		window: { showErrorMessage: jest.fn() },
		tracer: { log: jest.fn() },
	}) as unknown as Connection;

const createMockLogger = () =>
	({
		debug: jest.fn(),
		info: jest.fn(),
		warn: jest.fn(),
		error: jest.fn(),
	}) as unknown as winston.Logger;

const createMockTextDocument = (nonFileURI = false) =>
	({
		uri: nonFileURI ? 'scheme:///fake/cwd/document.css' : 'file:///fake/cwd/document.css',
	}) as TextDocument;

const goodStylelintPath = mockedPath.join(__dirname, 'stylelint.js');
const badStylelintPath = mockedPath.join(__dirname, 'bad-stylelint.js');

const pnpPath = mockedPath.join(__dirname, '.pnp.cjs');
const pnpJSPath = mockedPath.join(__dirname, '.pnp.js');

const mockGlobalPaths: { [packageManager in PackageManager]: string } = {
	yarn: mockedPath.join('/fake', 'yarn'),
	npm: mockedPath.join('/fake', 'npm'),
	pnpm: mockedPath.join('/fake', 'pnpm'),
};

jest.doMock(path.join(__dirname, 'stylelint.js'), () => ({ lint: jest.fn(() => 'good') }), {
	virtual: true,
});
jest.doMock(path.join(__dirname, 'bad-stylelint.js'), () => ({}), {
	virtual: true,
});
jest.doMock(path.join(__dirname, '.pnp.cjs'), () => ({ setup: jest.fn() }), {
	virtual: true,
});
jest.doMock(path.join(__dirname, '.pnp.js'), () => ({ setup: jest.fn() }), {
	virtual: true,
});

// eslint-disable-next-line @typescript-eslint/no-require-imports
const mockedPnP = require(pnpPath) as jest.Mocked<{ setup: () => void }>;
// eslint-disable-next-line @typescript-eslint/no-require-imports
const mockedJSPnP = require(pnpJSPath) as jest.Mocked<{ setup: () => void }>;

const mockGlobalFileResolution = (packageManager: PackageManager, stylelintPath: string) => {
	mockedFiles.__mockResolution('stylelint', (globalPath, cwd, trace) => {
		if (trace) trace('Resolving globally');

		return cwd === mockCWD && globalPath === mockGlobalPaths[packageManager]
			? stylelintPath
			: undefined;
	});
};

const mockLocalFileResolution = (stylelintPath: string) => {
	mockedFiles.__mockResolution('stylelint', (_, cwd, trace) => {
		if (trace) trace('Resolving locally');

		return cwd === mockCWD ? stylelintPath : undefined;
	});
};

mockedGlobalPathResolver.mockImplementation(
	() =>
		({
			resolve: jest.fn(async (packageManager: PackageManager) => mockGlobalPaths[packageManager]),
		}) as unknown as GlobalPathResolver,
);

describe('StylelintResolver', () => {
	beforeEach(() => {
		jest.clearAllMocks();
		mockedPath.__mockPlatform();
		mockCWD = mockedPath.join('/fake', 'cwd');
		mockPnPVersion = undefined;
		mockedFS.stat.mockReset();
		mockedFindPackageRoot.mockReset();
	});

	test('should resolve valid custom Stylelint paths', async () => {
		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve(
			{ stylelintPath: goodStylelintPath },
			createMockTextDocument(),
		);

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).not.toHaveBeenCalled();
	});

	test('should resolve valid relative custom Stylelint paths with a workspace', async () => {
		mockCWD = __dirname;

		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve(
			{ stylelintPath: './stylelint.js' },
			createMockTextDocument(),
		);

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).not.toHaveBeenCalled();
	});

	test('should resolve valid relative custom Stylelint paths without a workspace', async () => {
		mockCWD = undefined;
		mockedPath.isAbsolute.mockReturnValueOnce(false);

		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve(
			{ stylelintPath: goodStylelintPath },
			createMockTextDocument(),
		);

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).not.toHaveBeenCalled();
	});

	test('should resolve to undefined for custom Stylelint paths pointing to modules without a lint function', async () => {
		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve(
			{ stylelintPath: badStylelintPath },
			createMockTextDocument(),
		);

		expect(result).toBeUndefined();
		expect(logger.warn).toHaveBeenCalledTimes(2);
		expect(logger.error).toHaveBeenCalledTimes(1);
		expect(connection.window.showErrorMessage).toHaveBeenCalledTimes(1);
		expect(connection.tracer.log).not.toHaveBeenCalled();
	});

	test('should throw on invalid custom Stylelint paths', async () => {
		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);

		mockCWD = mockedPath.join('.', 'cwd');
		mockedPath.__mockPlatform('posix');

		await expect(
			stylelintResolver.resolve({ stylelintPath: './does-not-exist' }, createMockTextDocument()),
		).rejects.toThrowErrorMatchingSnapshot();
		expect(logger.error).toHaveBeenCalledTimes(1);
		expect(connection.window.showErrorMessage).toHaveBeenCalledTimes(1);
		expect(connection.tracer.log).not.toHaveBeenCalled();
	});

	test('should resolve workspace Stylelint modules', async () => {
		mockLocalFileResolution(goodStylelintPath);

		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).toHaveBeenCalledTimes(1);
	});

	test('should resolve workspace Stylelint modules for documents with non-file URIs', async () => {
		mockLocalFileResolution(goodStylelintPath);

		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve({}, createMockTextDocument(true));

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).toHaveBeenCalledTimes(1);
	});

	test('should resolve global Stylelint modules using yarn', async () => {
		mockGlobalFileResolution('yarn', goodStylelintPath);

		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve(
			{ packageManager: 'yarn' },
			createMockTextDocument(),
		);

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).toHaveBeenCalledTimes(1);
	});

	test('should resolve global Stylelint modules using npm', async () => {
		mockGlobalFileResolution('npm', goodStylelintPath);

		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve(
			{ packageManager: 'npm' },
			createMockTextDocument(),
		);

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).toHaveBeenCalledTimes(1);
	});

	test('should resolve global Stylelint modules using pnpm', async () => {
		mockGlobalFileResolution('pnpm', goodStylelintPath);

		const connection = createMockConnection();
		const stylelintResolver = new StylelintResolver(connection);
		const result = await stylelintResolver.resolve(
			{ packageManager: 'pnpm' },
			createMockTextDocument(),
		);

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');
		expect(connection.console.error).not.toHaveBeenCalled();
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).toHaveBeenCalledTimes(1);
	});

	test('should return undefined when no Stylelint module is found globally or in the workspace', async () => {
		mockedFiles.__resetMockedResolutions();

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(result).toBeUndefined();
		expect(logger.warn).toHaveBeenCalledTimes(2);
		expect(connection.window.showErrorMessage).not.toHaveBeenCalled();
		expect(connection.tracer.log).not.toHaveBeenCalled();
	});

	test('should resolve workspace Stylelint modules using PnP', async () => {
		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockResolvedValue(__dirname);
		mockedFS.stat.mockResolvedValueOnce({
			isFile: () => true,
		} as unknown as Stats);
		mockedModule.createRequire.mockImplementation(
			() =>
				Object.assign(() => ({ lint: () => 'from pnp' }), {
					resolve: () => goodStylelintPath,
				}) as unknown as NodeRequire,
		);

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedPnP.setup).toHaveBeenCalledTimes(1);
		expect(logger.debug).toHaveBeenCalledWith('Resolved Stylelint using PnP', {
			path: pnpPath,
		});
		expect(result?.resolvedPath).toBe(__dirname);
		expect(result?.stylelint?.lint({})).toBe('from pnp');
	});

	test('should resolve workspace Stylelint modules using a PnP loader named .pnp.js (Yarn 2)', async () => {
		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockResolvedValue(__dirname);
		mockedFS.stat.mockImplementation(async (filePath) => {
			if (filePath.toString().endsWith('.pnp.js')) {
				return { isFile: () => true } as unknown as Stats;
			}

			throw new Error('Not found!');
		});

		mockedModule.createRequire.mockImplementation(
			() =>
				Object.assign(() => ({ lint: () => 'from pnp' }), {
					resolve: () => goodStylelintPath,
				}) as unknown as NodeRequire,
		);

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedJSPnP.setup).toHaveBeenCalledTimes(1);
		expect(logger.debug).toHaveBeenCalledWith('Resolved Stylelint using PnP', {
			path: pnpJSPath,
		});
		expect(result?.resolvedPath).toBe(__dirname);
		expect(result?.stylelint?.lint({})).toBe('from pnp');
	});

	test('should not try to setup PnP if it is already setup', async () => {
		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockResolvedValue(__dirname);
		mockedFS.stat.mockResolvedValueOnce({
			isFile: () => true,
		} as unknown as Stats);
		mockedModule.createRequire.mockImplementation(
			() =>
				Object.assign(() => ({ lint: () => 'from pnp' }), {
					resolve: () => goodStylelintPath,
				}) as unknown as NodeRequire,
		);
		mockPnPVersion = '1.0.0';

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);

		await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedPnP.setup).toHaveBeenCalledTimes(0);
	});

	test("should resolve to undefined if PnP setup fails and Stylelint can't be resolved from node_modules", async () => {
		const error = new Error('PnP setup failed');

		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockResolvedValue(__dirname);
		mockedFS.stat.mockResolvedValueOnce({
			isFile: () => true,
		} as unknown as Stats);
		mockedPnP.setup.mockImplementationOnce(() => {
			throw error;
		});
		mockedModule.createRequire.mockImplementation(
			() =>
				Object.assign(() => ({ lint: () => 'from pnp' }), {
					resolve: () => goodStylelintPath,
				}) as unknown as NodeRequire,
		);

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedPnP.setup).toHaveBeenCalledTimes(1);
		expect(logger.warn).toHaveBeenCalledWith('Could not setup PnP', {
			path: pnpPath,
			error,
		});
		expect(result).toBeUndefined();
	});

	test("should resolve to undefined if PnP loader isn't a file and Stylelint can't be resolved from node_modules", async () => {
		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockResolvedValue(__dirname);
		mockedFS.stat.mockResolvedValueOnce({
			isFile: () => false,
		} as unknown as Stats);
		mockedModule.createRequire.mockImplementation(
			() =>
				Object.assign(() => ({ lint: () => 'from pnp' }), {
					resolve: () => goodStylelintPath,
				}) as unknown as NodeRequire,
		);

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedPnP.setup).not.toHaveBeenCalled();
		expect(logger.debug).toHaveBeenCalledWith('Could not find a PnP loader', {
			path: __dirname,
		});
		expect(result).toBeUndefined();
	});

	test("should resolve to undefined if PnP loader can't be found and Stylelint can't be resolved from node_modules", async () => {
		const error = new Error('EACCES');

		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockResolvedValue(__dirname);
		mockedFS.stat.mockRejectedValueOnce(error);
		mockedModule.createRequire.mockImplementation(
			() =>
				Object.assign(() => ({ lint: () => 'from pnp' }), {
					resolve: () => goodStylelintPath,
				}) as unknown as NodeRequire,
		);

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedPnP.setup).not.toHaveBeenCalled();
		expect(logger.debug).toHaveBeenCalledWith('Could not find a PnP loader', {
			path: __dirname,
		});
		expect(result).toBeUndefined();
	});

	test("should resolve to undefined if Stylelint can't be required using PnP", async () => {
		const error = new Error('Cannot find module');

		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockResolvedValue(__dirname);
		mockedFS.stat.mockResolvedValueOnce({
			isFile: () => true,
		} as unknown as Stats);
		mockedPnP.setup.mockImplementationOnce(() => {
			mockedModule.createRequire.mockImplementationOnce(
				() =>
					Object.assign(
						() => {
							throw error;
						},
						{
							resolve: () => {
								throw error;
							},
						},
					) as unknown as NodeRequire,
			);
		});

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedPnP.setup).toHaveBeenCalledTimes(1);
		expect(logger.warn).toHaveBeenCalledWith('Could not load Stylelint using PnP', {
			path: __dirname,
			error,
		});
		expect(result).toBeUndefined();
	});

	test("should resolve to undefined if Stylelint path can't be determined using PnP", async () => {
		mockCWD = mockedPath.join('/fake', 'pnp');
		mockedFindPackageRoot.mockImplementation((startPath) =>
			Promise.resolve(startPath === mockedPath.join('/fake', 'cwd') ? __dirname : undefined),
		);
		mockedFS.stat.mockResolvedValueOnce({
			isFile: () => true,
		} as unknown as Stats);
		mockedModule.createRequire.mockImplementation(
			() =>
				Object.assign(() => ({ lint: () => 'from pnp' }), {
					resolve: () => goodStylelintPath,
				}) as unknown as NodeRequire,
		);

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument());

		expect(mockedPnP.setup).toHaveBeenCalledTimes(1);
		expect(logger.warn).toHaveBeenCalledWith('Failed to find the Stylelint package root', {
			path: goodStylelintPath,
		});
		expect(result).toBeUndefined();
	});

	test('should resolve to undefined if an error is thrown during resolution', async () => {
		mockedFindPackageRoot.mockRejectedValueOnce(new Error());

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument(true));

		expect(result).toBeUndefined();
	});

	test("should resolve to undefined if cwd can't be determined and Stylelint can't be resolved from node_modules", async () => {
		mockCWD = undefined;

		const connection = createMockConnection();
		const logger = createMockLogger();
		const stylelintResolver = new StylelintResolver(connection, logger);
		const result = await stylelintResolver.resolve({}, createMockTextDocument(true));

		expect(result).toBeUndefined();
	});

	test('should work without a connection', async () => {
		mockGlobalFileResolution('npm', goodStylelintPath);

		let result = await new StylelintResolver().resolve(
			{ packageManager: 'npm' },
			createMockTextDocument(),
		);

		expect(result?.resolvedPath).toBe(goodStylelintPath);
		expect(result?.stylelint?.lint({})).toBe('good');

		mockGlobalFileResolution('npm', badStylelintPath);

		result = await new StylelintResolver().resolve(
			{ packageManager: 'npm' },
			createMockTextDocument(),
		);

		expect(result).toBeUndefined();

		await expect(
			new StylelintResolver().resolve(
				{ stylelintPath: './does-not-exist' },
				createMockTextDocument(),
			),
		).rejects.toThrowErrorMatchingSnapshot();
	});
});
