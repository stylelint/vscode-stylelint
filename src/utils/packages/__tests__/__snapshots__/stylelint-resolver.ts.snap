// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`StylelintResolver should throw on invalid custom Stylelint paths 1`] = `"Cannot find module 'cwd/does-not-exist' from 'src/utils/packages/stylelint-resolver.ts'"`;

exports[`StylelintResolver should work without a connection 1`] = `"Cannot find module './does-not-exist' from 'src/utils/packages/stylelint-resolver.ts'"`;
